### PR TITLE
Fix miri error in into_inner() of CString

### LINKED ISSUE #62553: CString::into_raw() trigger miri
```rust
use std::ffi::CString;

fn main() {
    let _hello = CString::new("Hello")
        .expect("CString::new failed")
        .into_raw();
}
```

This simple code should not trigger any error, except a leak of course. But miri report an error before:

```none
error[E0080]: Miri evaluation error: trying to reborrow for Unique, but parent tag <2145> does not have an appropriate item in the borrow stack
   --> /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/ffi/c_str.rs:605:13
    |
605 |             result
    |             ^^^^^^ Miri evaluation error: trying to reborrow for Unique, but parent tag <2145> does not have an appropriate item in the borrow stack
    |
    = note: inside call to `std::ffi::CString::into_inner` at /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/ffi/c_str.rs:440:23
note: inside call to `std::ffi::CString::into_raw` at src/main.rs:4:18
   --> src/main.rs:4:18
    |
4   |       let _hello = CString::new("Hello")
    |  __________________^
5   | |         .expect("CString::new failed")
6   | |         .into_raw();
    | |___________________^
    = note: inside call to `main` at /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:64:34
    = note: inside call to closure at /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:52:53
    = note: inside call to closure at /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/panicking.rs:294:40
    = note: inside call to `std::panicking::try::do_call::<[closure@DefId(1:5878 ~ std[82ff]::rt[0]::lang_start_internal[0]::{{closure}}[0]) 0:&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe], i32>` at /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/panicking.rs:290:5
    = note: inside call to `std::panicking::try::<i32, [closure@DefId(1:5878 ~ std[82ff]::rt[0]::lang_start_internal[0]::{{closure}}[0]) 0:&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe]>` at /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/panic.rs:388:9
    = note: inside call to `std::panic::catch_unwind::<[closure@DefId(1:5878 ~ std[82ff]::rt[0]::lang_start_internal[0]::{{closure}}[0]) 0:&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe], i32>` at /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:52:25
    = note: inside call to `std::rt::lang_start_internal` at /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:64:5
    = note: inside call to `std::rt::lang_start::<()>`
```

First, I suspected a miri bug but look like the code of `CString` could be the problem, I don't really understand the code of the [`into_inner()`](https://github.com/rust-lang/rust/blob/e7efdf1c33c699def0e594f03337efc78120bd9c/src/libstd/ffi/c_str.rs#L601) call by [`into_raw()`](https://github.com/rust-lang/rust/blob/e7efdf1c33c699def0e594f03337efc78120bd9c/src/libstd/ffi/c_str.rs#L439).

```rust
fn into_inner(self) -> Box<[u8]> {
    unsafe {
        let result = ptr::read(&self.inner);
        mem::forget(self);
        result
    }
}
```

Is this code correct and it's a miri bug or the code is incorrect ?

@matklad as you write the code maybe you want be ping.

### PR DESCRIPTION
**Pull request by Stargateur**
Fix #62553

I choice to not transmute because I think it's more unsafe and in case the structure change this code should always work.

r? @RalfJung

### CODE REVIEW

File: src/libstd/ffi/c_str.rs [hunk 1]
```diff
@@ -599,11 +599,12 @@ impl CString {
     ///
     /// [`Drop`]: ../ops/trait.Drop.html
     fn into_inner(self) -> Box<[u8]> {
-        unsafe {
-            let result = ptr::read(&self.inner);
-            mem::forget(self);
-            result
-        }
+        // Rational: `mem::forget(self);` invalid a previous call to `ptr::read(&self.inner)`
+        // so we use ManuallyDrop to unsure `self` is not drop
```
Comment (hellow554) [inline src/libstd/ffi/c_str.rs]: ensure

File: src/libstd/ffi/c_str.rs [hunk 2]
```diff
@@ -599,11 +599,12 @@ impl CString {
...
+        // and then we can return the box directly without invalid it.
```
Comment (hellow554) [inline src/libstd/ffi/c_str.rs]: invalidating

File: src/libstd/ffi/c_str.rs [hunk 3]
```diff
@@ -599,11 +599,12 @@ impl CString {
     ///
     /// [`Drop`]: ../ops/trait.Drop.html
     fn into_inner(self) -> Box<[u8]> {
-        unsafe {
-            let result = ptr::read(&self.inner);
-            mem::forget(self);
-            result
-        }
+        // Rational: `mem::forget(self);` invalid a previous call to `ptr::read(&self.inner)`
```
Comment (hellow554) [inline src/libstd/ffi/c_str.rs]: invalidates the

**Review by hellow554 — CHANGES_REQUESTED** [review]

**Comment by Stargateur** [comment]
I would like to add a test but I don't know how.

File: src/libstd/ffi/c_str.rs [hunk 4]
```diff
@@ -599,11 +599,12 @@ impl CString {
     ///
     /// [`Drop`]: ../ops/trait.Drop.html
     fn into_inner(self) -> Box<[u8]> {
-        unsafe {
-            let result = ptr::read(&self.inner);
-            mem::forget(self);
-            result
-        }
+        // Rational: `mem::forget(self);` invalidates the previous call to `ptr::read(&self.inner)`
+        // so we use ManuallyDrop to ensure `self` is not drop
```
Comment (RalfJung) [inline src/libstd/ffi/c_str.rs]:
```suggestion
        // so we use ManuallyDrop to ensure `self` is not dropped.
```

File: src/libstd/ffi/c_str.rs [hunk 5]
```diff
@@ -599,11 +599,12 @@ impl CString {
...
+        // and then we can return the box directly without invalidating it.
```
Comment (RalfJung) [inline src/libstd/ffi/c_str.rs]:
```suggestion
        // Then we can return the box directly without invalidating it.
```

File: src/libstd/ffi/c_str.rs [hunk 6]
```diff
@@ -599,11 +599,12 @@ impl CString {
...
+        // see https://github.com/rust-lang/rust/issues/62553
```
Comment (RalfJung) [inline src/libstd/ffi/c_str.rs]:
```suggestion
        // See https://github.com/rust-lang/rust/issues/62553.
```

File: src/libstd/ffi/c_str.rs [hunk 7]
```diff
@@ -599,11 +599,12 @@ impl CString {
     ///
     /// [`Drop`]: ../ops/trait.Drop.html
     fn into_inner(self) -> Box<[u8]> {
-        unsafe {
-            let result = ptr::read(&self.inner);
-            mem::forget(self);
-            result
-        }
+        // Rational: `mem::forget(self);` invalidates the previous call to `ptr::read(&self.inner)`
```
Comment (RalfJung) [inline src/libstd/ffi/c_str.rs]:
```suggestion
        // Rationale: `mem::forget(self)` invalidates the previous call to `ptr::read(&self.inner)`
```

**Comment by RalfJung** [comment]
> I would like to add a test but I don't know how.

You mean a test that this does not error in Miri any more? For that, please make a PR at I would like to add a test but I don't know how..

**Comment by RalfJung** [comment]
This looks great! When you force-push and want a new review, please leave a note saying so; GitHub unfortunately does not send email notifications for force-pushes.

@bors r+

Do you want to try to add a Miri test case as well? Do you need mentoring for that?

**Pull request merged**